s3transfer/upload.py

```diff
@@ -286,6 +286,157 @@ def _get_put_object_fileobj(self, fileobj):
+        # hanging around in memory.
+        self._initial_data = None
+        return body
+
+    def yield_upload_part_bodies(self, transfer_future, config):
+        part_size = config.multipart_chunksize
+        file_object = transfer_future.meta.call_args.fileobj
+        callbacks = get_callbacks(transfer_future, 'progress')
+        part_number = 0
+
+        # First go over the parts generated by calling
+        # `requires_multipart_upload` if any.
+        if self._initial_data is not None:
+            parts = self._yield_first_parts(transfer_future, config, callbacks)
+            for part in parts:
+                part_number += 1
+                yield part_number, part
+
+        # Continue reading parts from the file-like object until it is empty.
+        while True:
+            part_number += 1
+            part_content = file_object.read(part_size)
+            if not part_content:
+                break
+            part_object = self._wrap_part_data(part_content, callbacks)
+
+            # Zero out part_content to avoid hanging on to additional data.
+            part_content = None
+            yield part_number, part_object
+
+    def _yield_first_parts(self, transfer_future, config, callbacks):
+        """
+        Yields parts contained within the initial data.
+
+        :type transfer_future: s3transfer.futures.TransferFuture
+        :param transfer_future: The future associated with upload request.
+
+        :type config: s3transfer.manager.TransferConfig
+        :param config: The config associated to the transfer manager.
+
+        :type callbacks: list
+        :param callbacks: The callbacks associated with the transfer future.
+
+        :return: Generator which generates part bodies from the initial data.
+        """
+        try:
+            fileobj = transfer_future.meta.call_args.fileobj
+            part_size = config.multipart_chunksize
+
+            data_length = len(self._initial_data)
+
+            for i in range(0, data_length, part_size):
+                part_bound = i + part_size
+                if part_bound > data_length:
+                    # The remaining data is not enough for an entire part. We
+                    # now attempt to red additional data to fill out the part.
+                    data_to_read = part_bound - data_length
+                    yield self._wrap_part_data(
+                        data=self._initial_data[i:]+fileobj.read(data_to_read),
+                        callbacks=callbacks)
+                else:
+                    yield self._wrap_part_data(
+                        self._initial_data[i:part_bound], callbacks)
+        except GeneratorExit as e:
+            # Zero out the initial data now that we're done generating parts
+            # from it so that we don't hold onto those bytes any longer
+            # than we need to.
+            self._initial_data = None
+            raise e
+
+    def _wrap_part_data(self, data, callbacks, size=None):
```

Comment:
I think we can call it something more general like ``_wrap_data`` as this is used for PutObject and UploadPart.